ddl/db_partition_test.go

```diff
@@ -771,6 +771,286 @@ func (s *testIntegrationSuite5) TestAlterTableDropPartition(c *C) {
+		},
+		{
+			"create table pt20 (id int not null) partition by hash(id) partitions 1;",
+			"create table nt20 (id int default null);",
+			"alter table pt20 exchange partition p0 with table nt20;",
+			ddl.ErrTablesDifferentMetadata,
+		},
+		{
+			// unsigned
+			"create table pt21 (id int unsigned) partition by hash(id) partitions 1;",
+			"create table nt21 (id int);",
+			"alter table pt21 exchange partition p0 with table nt21;",
+			ddl.ErrTablesDifferentMetadata,
+		},
+		{
+			// zerofill
+			"create table pt22 (id int) partition by hash(id) partitions 1;",
+			"create table nt22 (id int zerofill);",
+			"alter table pt22 exchange partition p0 with table nt22;",
+			ddl.ErrTablesDifferentMetadata,
+		},
+		{
+			"create table pt23 (id int, lname varchar(10) charset binary) partition by hash(id) partitions 1;",
+			"create table nt23 (id int, lname varchar(10));",
+			"alter table pt23 exchange partition p0 with table nt23;",
+			ddl.ErrTablesDifferentMetadata,
+		},
+		{
+			"create table pt25 (id int, a datetime on update current_timestamp) partition by hash(id) partitions 1;",
+			"create table nt25 (id int, a datetime);",
+			"alter table pt25 exchange partition p0 with table nt25;",
+			nil,
+		},
+		{
+			"create table pt26 (id int not null, lname varchar(30), fname varchar(100) generated always as (concat(lname, ' ')) virtual) partition by hash(id) partitions 1;",
+			"create table nt26 (id int not null, lname varchar(30), fname varchar(100) generated always as (concat(id, ' ')) virtual);",
+			"alter table pt26 exchange partition p0 with table nt26;",
+			nil,
+		},
+	}
+
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	for i, t := range cases {
+		tk.MustExec(t.ptSQL)
+		tk.MustExec(t.ntSQL)
+		if t.err != nil {
+			_, err := tk.Exec(t.exchangeSQL)
+			originErr := errors.Cause(err)
+			tErr, ok := originErr.(*terror.Error)
+			c.Assert(ok, IsTrue, Commentf("case %d fail, sql = `%s`\nexpect type 'terror.Error', but obtain '%T'", i,
+				t.exchangeSQL, originErr))
+			sqlErr := tErr.ToSQLError()
+			c.Assert(int(t.err.Code()), Equals, int(sqlErr.Code), Commentf(
+				"case %d fail, sql = `%s`\nexpected error = `%v`\n  actual error = `%v`",
+				i, t.exchangeSQL, t.err, err,
+			))
+		} else {
+			tk.MustExec(t.exchangeSQL)
+		}
+	}
+
+}
+
+func (s *testIntegrationSuite7) TestExchangePartitionExpressIndex(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists pt1;")
+	tk.MustExec("create table pt1(a int, b int, c int) PARTITION BY hash (a) partitions 1;")
+	tk.MustExec("alter table pt1 add index idx((a+c));")
+
+	tk.MustExec("drop table if exists nt1;")
+	tk.MustExec("create table nt1(a int, b int, c int);")
+	tk.MustGetErrCode("alter table pt1 exchange partition p0 with table nt1;", tmysql.ErrTablesDifferentMetadata)
+
+	tk.MustExec("alter table nt1 add column (`_V$_idx_0` bigint(20) generated always as (a+b) virtual);")
```

Comment:
We cannot add expression index by this way, the right way is:
execute `config.GetGlobalConfig().Experimental.AllowsExpressionIndex = true` before `tk := testkit.NewTestKit(c, s.store)`, and then `alter table nt1 add index ((a+b))` directly.

Reply:
The `testIntegrationSuite7` is specifically used to test expression index. The code you mentioned have 
written in [db_integration_test.go](https://github.com/pingcap/tidb/blob/2efab88a59499ea3934c8c1eba7b695cf24b9d29/ddl/db_integration_test.go#L2051). Should I explicit declaration it?  So I confused what is the difference between `testIntegrationSuite 1-7 ` in ddl

Reply:
`testIntegrationSuite 1-7 ` is split by one test suit, why split the origin test suit is for speed up CI.
If you want a test case enable some feature, you should explicit enable it in the test case by:
```golang
config.GetGlobalConfig().Experimental.AllowsExpressionIndex = true
```
Even if a test case before may have enabled it, enable it again here can ensure this feature is definitely enabled.

Reply:
By the way, replace
```golang
alter table nt1 add column (`_V$_idx_0` bigint(20) generated always as (a+b) virtual);
```
with (the right way for create an expression index)
```golang
alter table nt1 add index ((a+b))
```

Reply:
While a expression index could auto generate a hidden virtual column, I simple regard it as a test case of exchange partition. The acutal expression index test is among line `1141-1128` while both tables hava a expression index.